### PR TITLE
Fix W5500 deadlock after processing interrupt to slow

### LINKED ISSUE #45735: Ethernet W5500 Driver via SPI is deadlocking
**Describe the bug**
When utilizing the W5500 for Ethernet via SPI with [Zephyr's driver](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/1.9.1/zephyr/reference/devicetree/bindings/ethernet/wiznet,w5500.html?highlight=wiznet), it will eventually become unresponsive due to the module no longer triggering the GPIO callback, which in turn never gives the "int" semaphore back which prevents it from running its `w5500_isr` interrupt function.

I'm using:
- Nordic nRF52840
- nRF Connect SDK 1.9.1 (Zephyr 2.7.1)
- SEGGER Embedded Studio / nRF Connect for VSCode
- I've checked Discord, and at least one other person came across this
- I've tried dropping in the recent driver and it gave me the same result. Maybe I'm missing configuration and/or docs are missing additional info?
- I've attempted increasing timeout, thread priority, RX stack size to no avail
- I've noticed SPI API for Zephyr is very slow compared to baremetal from another application I've used

**To Reproduce**
0. Get a board with the W5500 Ethernet module driver
1. Configure DHCPv4 or static IP Address
2. Configure Ethernet with the W5500 module driver
3. Build and Flash the board with the IDE/West/Ninja commands of your choice
4. Ping the board in your local network 
5. Observe that the response time is roughly 1.8ms (very long for local)
6. Eventually (40 mins) the device's ethernet interface will lock up, making the address unreachable and isolating the device.

This lock up can be sped up by using more demanding examples, like MQTT Publisher.

**Expected behavior**
Ethernet interface W5500 to run indefinitely after being set up.
If the module locks up, it should probably detect the lock and reset just the module with an error log.

**Impact**
This is a showstopper because the board must be reboot manually for the application to work upon detecting this fault, which is not possible to work with.

**Logs and console output**
Only logs I have are custom traces that lead to showing the `int_sem` no longer is given up, locking up the driver with no countermeasure.

The following no longer calls back after some time:
https://github.com/zephyrproject-rtos/zephyr/blob/734c3454578d4db89b7a879b7339559b20c04c44/drivers/ethernet/eth_w5500.c#L492-L493
Which leads to the following never calling:
https://github.com/zephyrproject-rtos/zephyr/blob/734c3454578d4db89b7a879b7339559b20c04c44/drivers/ethernet/eth_w5500.c#L423-L431
Which locks here:
https://github.com/zephyrproject-rtos/zephyr/blob/734c3454578d4db89b7a879b7339559b20c04c44/drivers/ethernet/eth_w5500.c#L283-L292
**Environment (please complete the following information):**
 - Linux Ubuntu Focal Fossa flashing a nRF52840 with JLink
 - gcc-arm-none-eabi-9-2019-q4-major with nRF Connect SDK 1.9.1
 - Nordic ncs [2.7.99](https://github.com/nrfconnect/sdk-zephyr/tree/v2.7.99-ncs1-branch)

**Additional context**
My board is custom so the SPI is set up as follows:
```js
&spi0 {
	compatible = "nordic,nrf-spim";
	sck-pin = <32>;
	miso-pin = <25>;
	mosi-pin = <24>;
	cs-gpios = <&gpio0 22 GPIO_ACTIVE_LOW>;
	status = "okay";
	w5500: w5500@0 {
		compatible = "wiznet,w5500";
		label = "w5500";
		reg = <0>;
		spi-max-frequency = <8000000>;
		int-gpios = <&gpio0 14 GPIO_ACTIVE_LOW>;
		reset-gpios = <&gpio0 13 GPIO_ACTIVE_LOW>;
	};
};
```

Any tips on how to fix this? I might make a PR myself.
Currently I don't understand why the GPIO stops, but I'm assuming it's because the SPI is too slow because after I increased the max frequency the problem took longer to occur (it's still too slow compared to bare metal though).

### PR DESCRIPTION
**Pull request by jancoow**
When the W5500 gives an interrupt, a semaphore will be given which will be handled inside a thread.

However, in the time that the interrupt will be cleared and the thread will be waiting for a new semaphore, the w5500 could already have a new interrupt available which will keep the INT pin low. This is not checked correctly, which will lead in a deadlock, which can be seen with a logic analyzer on the INT pin:

Screenshot-1

I solved this by always checking if the pin gets high after handling and clearing the interrupts. If not, there are more interrupts available and they should be handled immediately. 

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/45735

### CODE REVIEW

File: drivers/ethernet/eth_w5500.c [hunk 1]
```diff
@@ -208,7 +209,7 @@ static int w5500_tx(const struct device *dev, struct net_pkt *pkt)
 
 static void w5500_rx(const struct device *dev)
 {
-	uint8_t mask = 0;
+	// uint8_t mask = 0;
```
Comment (carlescufi) [inline drivers/ethernet/eth_w5500.c:212]: Do not leave commented out code please.

File: drivers/ethernet/eth_w5500.c [hunk 2]
```diff
@@ -222,7 +223,7 @@ static void w5500_rx(const struct device *dev)
 	const struct w5500_config *config = dev->config;
 
 	/* disable interrupt */
-	w5500_spi_write(dev, W5500_SIMR, &mask, 1);
+	// w5500_spi_write(dev, W5500_SIMR, &mask, 1);
```
Comment (carlescufi) [inline drivers/ethernet/eth_w5500.c:226]: Do not leave commented out code please.

File: drivers/ethernet/eth_w5500.c [hunk 3]
```diff
@@ -280,34 +281,37 @@ static void w5500_rx(const struct device *dev)
...
-		if (ir & S0_IR_SENDOK) {
-			k_sem_give(&ctx->tx_sem);
-			LOG_DBG("TX Done");
-		}
-
-		if (ir & S0_IR_RECV) {
-			w5500_rx(dev);
+				
+		while(gpio_pin_get_dt(&(config->interrupt))){
+			// Read interrupt
```
Comment (carlescufi) [inline drivers/ethernet/eth_w5500.c:295]: No C++-style comments

File: drivers/ethernet/eth_w5500.c [hunk 4]
```diff
@@ -280,34 +281,37 @@ static void w5500_rx(const struct device *dev)
...
+			w5500_spi_read(dev, W5500_S0_IR, &ir, 1);
+			
+			if (ir) {
+				// Clear interrupt
```
Comment (carlescufi) [inline drivers/ethernet/eth_w5500.c:299]: No C++-style comments

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: drivers/ethernet/eth_w5500.c (hunk 3 above)
Comment (rlubos) [inline drivers/ethernet/eth_w5500.c:295]: C-style comments only please (`/* */ `).

**Review by rlubos — COMMENTED** [review]
Please check the compliance results and address all warnings/errors.

File: drivers/ethernet/eth_w5500.c [hunk 5]
```diff
@@ -34,6 +34,7 @@ LOG_MODULE_REGISTER(eth_w5500, CONFIG_ETHERNET_LOG_LEVEL);
 #define W5500_SPI_WRITE_CONTROL(addr)   \
```
Comment (jukkar) [inline drivers/ethernet/eth_w5500.c:34]: Please change the commit subject to
`drivers: ethernet: w5500: Check int pin after processing interrupt`
Also there is line length limits in commit message body so you should split it to multiple lines (sanity checker will complain about those)

**Comment by GeorgeCGV** [comment]
Doesn't it indicate a bigger problem?

`k_sem_give`

> A semaphore may be given by a thread or an ISR. Giving the semaphore increments its count unless the count is already equal to the limit.

and that is being done in the GPIO callback that is supposed to be called by the interrupt. That shall be called when `w5500_isr` thread runs and increment the count of the `int_sem` that is initialized with `UINT_MAX` limit.

As soon as `w5500_isr` thread is finished with package processing it shall be able to perform a successful `k_sem_take` of the `int_sem`. As it is going to be given by the GPIO callback.

Even if there is a huge amount of requests they shall still increment the semaphore limit that shall allow the `w5500_isr` thread to take the semaphore as soon as it can.

Considering that the issue exists means that there is some kind of a problem with the semaphore or GPIO callback handling, isn't it?

**Comment by jancoow** [comment]
No, the INT pin is kept low by the W5500 so a new interrupted is never fired.

**Comment by GeorgeCGV** [comment]
> No, the INT pin is kept low by the W5500 so a new interrupted is never fired.

Doesn't it mean that the interrupt is not cleared on time?

**Comment by jancoow** [comment]
The interrupts are cleared immediately after reading them, however an other interrupt could occur between those two calls which can keep the pin low.

**Comment by GeorgeCGV** [comment]
@jancoow, but when it is cleared the pin shall go H and then L again or?

Maybe `INTLEVEL (Interrupt Low Level Timer Register) [R/W]` is a solution?

> INTLEVEL configures the Interrupt Assert Wait Time (IAWT). When the next interrupt
occurs, Interrupt PIN (INTn ) will assert to low after INTLEVEL time.

**Comment by jancoow** [comment]
We are not clearing all interrupts. We are clearing the interrupts read from the register. Between reading and  clearing the ones in the register, a different one could always occur. So this check is always needed for safety anyways, as far as I can tell.

**Comment by lucasgcbkhomp** [comment]
@jukkar May we get the workflows for this to run? The results of this PR would be interesting to investigate further.

**Review by parthitce — APPROVED** [review]
Thanks for fixing this. LGTM.

**Comment by jancoow** [comment]
> LGTM

Took me a _few_ attempts to get it through all the checks

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
@carlescufi please revisit

**Review by carlescufi — APPROVED** [review]

**Comment by LukeShu** [comment]
Just as a note to anyone else who comes across this when troubleshooting their own W5500 code:

By the datasheet (v1.1.0), setting INTLEVEL should have solved this. It specifically describes this scenario on page 34: If interrupt A fires and pulls INT low, then interrupt B fires, then the host clears interrupt A without clearing interrupt B; then clearing A will pull INT back high for the INTLEVEL-configured time, then it will go back low because B hasn't been cleared yet. And so the semaphore would get incremented again, albeit after a delay.

But the datasheet lies; in reality INT does not go high until all the interrupts are cleared.

**Pull request merged**